grid-proxy/internal/indexer/dmi.go

```diff
@@ -0,0 +1,176 @@
+)
+
+const (
+	DmiCallCmd       = "zos.system.dmi"
+	flushingInterval = 60 * time.Second
+	dmiBatchSize     = 20
+)
+
+type DmiWatcher struct {
+	database        db.Database
+	rmbClient       *peer.RpcClient
+	nodeTwinIdsChan chan uint32
+	resultChan      chan types.DmiInfo
+	interval        time.Duration
+	workers         uint
+	batchSize       uint
+}
+
+func NewDmiWatcher(
+	ctx context.Context,
+	database db.Database,
+	rmbClient *peer.RpcClient,
+	interval uint,
+	workers uint,
+) *DmiWatcher {
+	return &DmiWatcher{
+		database:        database,
+		rmbClient:       rmbClient,
+		nodeTwinIdsChan: make(chan uint32),
+		resultChan:      make(chan types.DmiInfo),
+		interval:        time.Duration(interval) * time.Minute,
+		workers:         workers,
+		batchSize:       dmiBatchSize,
+	}
+}
+
+func (w *DmiWatcher) Start(ctx context.Context) {
+	go w.startNodeQuerier(ctx)
+
+	for i := uint(0); i < w.workers; i++ {
+		go w.startNodeCaller(ctx)
+	}
+
+	go w.startUpserter(ctx, w.database)
+}
+
+// TODO: not only on interval but also on any node goes from down>up or newly added nodes
+func (w *DmiWatcher) startNodeQuerier(ctx context.Context) {
+	ticker := time.NewTicker(w.interval)
+	queryUpNodes(ctx, w.database, w.nodeTwinIdsChan)
+	for {
+		select {
+		case <-ticker.C:
+			queryUpNodes(ctx, w.database, w.nodeTwinIdsChan)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (w *DmiWatcher) startNodeCaller(ctx context.Context) {
+	for {
+		select {
+		case twinId := <-w.nodeTwinIdsChan:
+			response, err := w.callNode(ctx, twinId)
+			if err != nil {
+				continue
+			}
+			parsedDmi := parseDmiResponse(response)
+			parsedDmi.NodeTwinId = twinId
+			w.resultChan <- parsedDmi
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// TODO: make it generic and then assert the result in each watcher
+func (w *DmiWatcher) callNode(ctx context.Context, twinId uint32) (DMI, error) {
+	var result DMI
+	subCtx, cancel := context.WithTimeout(ctx, indexerCallTimeout)
+	defer cancel()
+
+	err := w.rmbClient.Call(subCtx, twinId, DmiCallCmd, nil, &result)
+	if err != nil {
+		log.Error().Err(err).Uint32("twinId", twinId).Msg("failed to call node")
+	}
+
+	return result, err
+}
+
+func (w *DmiWatcher) startUpserter(ctx context.Context, database db.Database) {
+	buffer := make([]types.DmiInfo, 0, w.batchSize)
+
+	ticker := time.NewTicker(flushingInterval)
+	for {
+		select {
+		case dmiData := <-w.resultChan:
+			buffer = append(buffer, dmiData)
+			if len(buffer) >= int(w.batchSize) {
+				err := w.database.UpsertNodeDmi(ctx, buffer)
+				if err != nil {
+					log.Error().Err(err).Msgf("failed")
+				}
+				buffer = nil
+			}
+		case <-ticker.C:
+			if len(buffer) != 0 {
+				err := w.database.UpsertNodeDmi(ctx, buffer)
+				if err != nil {
+					log.Error().Err(err).Msgf("failed")
+				}
+				buffer = nil
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func parseDmiResponse(dmiResponse DMI) types.DmiInfo {
```

Comment:
replace everything related to parsing by zos dmidecode package, don't do it yourself

Reply:
checked the dmi pkg in zos and the `DMI` response is already parsed from the pkg in zos.
what i am doing here is just selecting what we want to store instead of storing the whole dmi response